checks.d/openstack.py

```diff
@@ -0,0 +1,827 @@
+                set(self.get_servers_managed_by_hypervisor()) - set(excluded_server_ids)
+            )
+            host_tags = self._get_tags_for_host()
+
+            for sid in servers:
+                server_tags = ['server:%s' % sid, "nova_managed_server"] + host_tags
+                self.external_host_tags[sid] = host_tags
+                self.get_stats_for_single_server(sid, tags=server_tags)
+
+            if hyp:
+                self.get_stats_for_single_hypervisor(hyp, host_tags=host_tags)
+            else:
+                self.warning("Couldn't get hypervisor to monitor for host: %s" % self.get_my_hostname())
+
+            if project:
+                self.get_stats_for_single_project(project)
+
+            # For now, monitor all networks
+            self.get_network_stats()
+
+        except MissingEndpoint as e:
+            if isinstance(e, MissingNeutronEndpoint):
+                self.warning("The agent could not find a compatible Neutron endpoint in your service catalog!")
+                self.service_check(self.NETWORK_API_SC, AgentCheck.CRITICAL, tags=["keystone_server:%s" % self.init_config.get("keystone_server_url")])
+            elif isinstance(e, MissingNovaEndpoint):
+                self.warning("The agent could not find a compatible Nova endpoint in your service catalog!")
+                self.service_check(self.COMPUTE_API_SC, AgentCheck.CRITICAL, tags=["keystone_server:%s" % self.init_config.get("keystone_server_url")])
+            else:
+                self.warning("Missing endpoint in service catalog: %s" % e)
+
+        except KeystoneUnreachable:
+            self.warning("The agent could not contact the specified identity server at %s . Are you sure it is up at that address?" % self.init_config.get("keystone_server_url"))
+            self.service_check(self.IDENTITY_API_SC, AgentCheck.CRITICAL, tags=["server:%s" % self.init_config.get("keystone_server_url")])
+
+        except IncompleteConfig as e:
+            if isinstance(e, IncompleteAuthScope):
+                self.warning("""Please specify the auth scope via the `auth_scope` variable in your init_config.\n
+                             The auth_scope should look like: \n
+                            {'project': {'name': 'my_project', 'domain': {'id': 'my_domain_id'}}}\n
+                            OR\n
+                            {'project': {'id': 'my_project_id'}}
+                             """)
+            elif isinstance(e, IncompleteIdentity):
+                self.warning("Please specify the user via the `user` variable in your init_config.\n" +
+                             "This is the user you would use to authenticate with Keystone v3 via password auth.\n" +
+                             "The user should look like: {'password': 'my_password', 'name': 'my_name', 'domain': {'id': 'my_domain_id'}}")
+            else:
+                self.warning("Configuration Incomplete! Check your openstack.yaml file")
+
+        else:
```

Comment:
I think this overall logic will result in cases where not all 3 service checks will send a result.
e.g. if KeystoneUnreachable is reached, only the IDENTIFY_API service check will send a status.